### PR TITLE
feat: Expose getIamPolicy support for project

### LINKED ISSUE #281: Add getIamPolicy() support for project
Request to implement `getIamPolicy()` call for the project object.

### PR DESCRIPTION
**Pull request by laljikanjareeya**
Expose [getIamPolicy](https://cloud.google.com/resource-manager/reference/rest/v1/projects/getIamPolicy) for project.
Fixes #281

### CODE REVIEW

File: src/project.ts [hunk 1]
```diff
@@ -332,6 +364,58 @@ class Project extends ServiceObject {
...
+   * @returns {Promise<GetIamPolicyResponse>}
+   *
+   * @example
+   * const {Resource} = require('@google-cloud/resource');
+   * const resource = new Resource();
+   * const project = resource.project('grape-spaceship-123');
+   *
+   * project.getIamPolicy((err, policy) => {
+   *   if (!err) {
+   *     // console.log(policy).
```
Comment (stephenplusplus) [inline src/project.ts]: You could probably do this as just:

```js
  *   if (!err) {
  *     console.log(policy);
  *   }
```

File: src/project.ts [hunk 2]
```diff
@@ -332,6 +364,58 @@ class Project extends ServiceObject {
...
+   */
+  /**
+   * @callback GetIamPolicyCallback
+   * @param {?Error} err Request error, if any.
+   * @param {Policy} policy The new [policy]{@link https://cloud.google.com/resource-manager/reference/rest/Shared.Types/Policy}.
+   */
+  /**
+   * Get Iam Policy for Project.
+   *
+   * @see [projects: getIamPolicy API Documentation]{@link https://cloud.google.com/resource-manager/reference/rest/v1/projects/getIamPolicy}
```
Comment (stephenplusplus) [inline src/project.ts]: This API accepts options. Should we add support here?

File: src/project.ts [hunk 3]
```diff
@@ -332,6 +364,58 @@ class Project extends ServiceObject {
       }
     );
   }
+
+  getIamPolicy(): Promise<GetIamPolicyResponse>;
+  getIamPolicy(callback: GetIamPolicyCallback): void;
+  /**
+   * @typedef {array} GetIamPolicyResponse
+   * @property {Policy} 0 The new [policy]{@link https://cloud.google.com/resource-manager/reference/rest/Shared.Types/Policy}.
```
Comment (stephenplusplus) [inline src/project.ts]: I don't know that we would say it's a "new" policy. Maybe just "This project's IAM policy."?

File: src/project.ts [hunk 4]
```diff
@@ -332,6 +364,58 @@ class Project extends ServiceObject {
...
+   */
+  /**
+   * @callback GetIamPolicyCallback
+   * @param {?Error} err Request error, if any.
+   * @param {Policy} policy The new [policy]{@link https://cloud.google.com/resource-manager/reference/rest/Shared.Types/Policy}.
```
Comment (stephenplusplus) [inline src/project.ts]: Same thought about the wording here.

File: src/project.ts [hunk 5]
```diff
@@ -332,6 +364,58 @@ class Project extends ServiceObject {
...
+   */
+  /**
+   * Get Iam Policy for Project.
```
Comment (stephenplusplus) [inline src/project.ts]: A little more friendly, maybe "Get the IAM policy for this project."

File: system-test/resource.ts [hunk 6]
```diff
@@ -57,6 +57,15 @@ describe('Resource', () => {
       const [metadata] = await project.getMetadata();
       assert.notStrictEqual(metadata.projectId, undefined);
     });
+
+    it('should get Iam policy', done => {
+      project.getIamPolicy((err, policy) => {
+        assert.ifError(err);
+        assert.notStrictEqual(policy!.etag, undefined);
+        assert.equal(typeof policy!.version, 'number');
```
Comment (stephenplusplus) [inline system-test/resource.ts]: `strictEqual` please!

File: test/project.ts [hunk 7]
```diff
@@ -142,4 +143,60 @@ describe('Project', () => {
...
+
+    it('should execute the callback with error & API response', done => {
+      project.getIamPolicy((err: Error, apiResponse_: Policy) => {
+        assert.strictEqual(err, error);
+        assert.strictEqual(apiResponse_, policy);
+        done();
+      });
+    });
+
+    it('should not require a callback', () => {
```
Comment (stephenplusplus) [inline test/project.ts]: I think we should require a callback, otherwise there's not a point to calling the method. One method that doesn't require a callback, like "restore()", the user could ignore the response, but the actions caused by the API request would still persist.

File: src/project.ts [hunk 8]
```diff
@@ -332,6 +364,58 @@ class Project extends ServiceObject {
       }
     );
   }
+
+  getIamPolicy(): Promise<GetIamPolicyResponse>;
```
Comment (stephenplusplus) [inline src/project.ts]: Could you sort this method alphabetically within the method declarations? It makes it easier to navigate the file when jumping in for a quick fix / investigation down the road.

File: test/project.ts [hunk 9]
```diff
@@ -142,4 +143,60 @@ describe('Project', () => {
       });
     });
   });
+
+  describe('getIamPolicy', () => {
```
Comment (stephenplusplus) [inline test/project.ts]: Could you also sort this alphabetically within the file?

File: test/project.ts (hunk 7 above)
Comment (laljikanjareeya) [reply test/project.ts]: Making callback require user will not able to use `await` here What do you think?

Comment (stephenplusplus) [reply test/project.ts]: Could you explain that further? If it relates, all of our methods require a callback that [`promisifyAll`](https://github.com/googleapis/nodejs-promisify/blob/592e063136ed296750be8740150025d171bdcc93/src/index.ts#L134) uses to convert the callback-written function into a promise-based function, that is then compatible with an async workflow.

File: src/project.ts (hunk 2 above)
Comment (laljikanjareeya) [reply src/project.ts]: Done!

File: test/project.ts (hunk 7 above)
Comment (laljikanjareeya) [reply test/project.ts]: Yes we are using `promisifyAll` to convert promise-based function.
I have seen in all repo we are doing same pattern, let me know if you have working example/reference of async workflow will make change accordingly.
Thanks!!

Comment (stephenplusplus) [reply test/project.ts]: The only thing I'm pointing out in this case, is that the callback should not be optional. Here's a comparable function: nodejs-compute/vm#attachDisk() https://github.com/googleapis/nodejs-compute/blob/2484e4b6521fdc9dce0c80b6e8bb4f3da7baf7df/src/vm.js#316

It accepts an optional `options` argument, and requires a callback.

All of these are compatible with a user who doesn't actually provide a function callback, instead using the promise/async flow.

**Comment by stephenplusplus** [comment]
@laljikanjareeya thank you for the changes!

File: test/project.ts (hunk 7 above)
Comment (AVaksman) [reply test/project.ts]: It seems that `nodejs-compute` is still `js` and, once converted and annotate with `typescript` types, the `callback` will have to be marked as optional as well. From user's perspective, `callback` is optional. It will nevertheless always be there where as from user or from `primisify`.

This existing test `it('should not require a callback', () => {...});` doesn't really test the functionality of this library, but rathe functionality of `promisify` and might not be needed here at all. WDYT?

Comment (stephenplusplus) [reply test/project.ts]: This is definitely hard to follow between all of our layers 👨💥 

The test is making sure that "callback = callback || util.noop" is there, which is what I think should not be there. I don't know if it makes a difference after promisify overwrites the method?, but if not, we could just assume there is always a callback, and remove the "defaulting to noop" line, and remove the test?

Comment (laljikanjareeya) [reply test/project.ts]: removed callback = callback || util.noop statement and test that was checking the callback.
